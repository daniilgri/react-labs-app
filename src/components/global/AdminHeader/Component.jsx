import React, { useState, useEffect, useRef } from "react";
import { faBars } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import {
  Wrapper,
  BackLink,
  MenuItem,
  Menu,
  StyledNavLink,
  MenuToggle,
} from "./styles";

const Component = () => {
  const [activeMenu, setActiveMenu] = useState(false);
  const [activeProfileSelect, setActiveProfileSelect] = useState(false);

  const profileSelectRef = useRef(null);

  const handleMenuToggleOnClick = (event) => {
    setActiveMenu((prevState) => !prevState);
  };

  const handleClickOutside = (event) => {
    if (
      activeProfileSelect &&
      profileSelectRef.current &&
      !profileSelectRef.current.contains(event.target)
    ) {
      setActiveProfileSelect(false);
    }
  };

  useEffect(() => {
    document.addEventListener("mousedown", handleClickOutside);
    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  });

  return (
    <Wrapper role="custom-dropdown">
      <MenuToggle onClick={handleMenuToggleOnClick}>
        <FontAwesomeIcon icon={faBars} />
      </MenuToggle>
      <BackLink to="/profile">Go to profile</BackLink>
      <Menu active={activeMenu}>
        <MenuItem>
          <StyledNavLink to="/admin/users">Users</StyledNavLink>
        </MenuItem>
        <MenuItem>
          <StyledNavLink exact to="/admin/films">
            Films
          </StyledNavLink>
        </MenuItem>
        <MenuItem>
          <StyledNavLink to="/admin/films/new">New film</StyledNavLink>
        </MenuItem>
      </Menu>
    </Wrapper>
  );
};

export default Component;
