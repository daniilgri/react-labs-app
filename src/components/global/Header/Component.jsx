import React, { useState, useEffect, useRef } from "react";
import PropTypes from "prop-types";
import { faBars, faUserCircle } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import {
  Wrapper,
  LogoWrapper,
  MenuItem,
  Menu,
  StyledNavLink,
  MenuToggle,
  Select,
  SelectIcon,
  OptionsList,
  Option,
  StyledLink,
  OptionLink,
} from "./styles";

const Component = ({ user, loading, error, signOutRequested }) => {
  const [activeMenu, setActiveMenu] = useState(false);
  const [activeProfileSelect, setActiveProfileSelect] = useState(false);

  const profileSelectRef = useRef(null);

  const handleMenuToggleOnClick = (event) => {
    setActiveMenu((prevState) => !prevState);
  };

  const handleProfileSelectOnClick = (event) => {
    setActiveProfileSelect((prevState) => !prevState);
  };

  const handleClickOutside = (event) => {
    if (
      activeProfileSelect &&
      profileSelectRef.current &&
      !profileSelectRef.current.contains(event.target)
    ) {
      setActiveProfileSelect(false);
    }
  };

  const handleDefaultCloseOptionOnClick = (event) => {
    setActiveProfileSelect(false);
  };

  const handleSignOutOptionOnClick = (event) => {
    event.preventDefault();
    signOutRequested();
    setActiveProfileSelect(false);
  };

  useEffect(() => {
    document.addEventListener("mousedown", handleClickOutside);
    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  });

  return (
    <Wrapper role="custom-dropdown">
      <MenuToggle onClick={handleMenuToggleOnClick}>
        <FontAwesomeIcon icon={faBars} />
      </MenuToggle>
      <LogoWrapper>Logo</LogoWrapper>
      <Menu active={activeMenu}>
        <MenuItem>
          <StyledNavLink to="/">All films</StyledNavLink>
        </MenuItem>
      </Menu>
      {!loading && (
        <Select active={activeMenu} ref={profileSelectRef}>
          <SelectIcon onClick={handleProfileSelectOnClick}>
            <FontAwesomeIcon icon={faUserCircle} />
          </SelectIcon>
          <OptionsList active={activeProfileSelect}>
            {user ? (
              <React.Fragment>
                <OptionLink
                  to="/profile"
                  onClick={handleDefaultCloseOptionOnClick}
                >
                  Profile
                </OptionLink>
                <OptionLink
                  to="/profile/orders"
                  onClick={handleDefaultCloseOptionOnClick}
                >
                  Orders
                </OptionLink>
                <Option onClick={handleSignOutOptionOnClick}>Sign out</Option>
              </React.Fragment>
            ) : (
              <React.Fragment>
                <OptionLink
                  to="/login"
                  onClick={handleDefaultCloseOptionOnClick}
                >
                  Login
                </OptionLink>
                <OptionLink
                  to="/signup"
                  onClick={handleDefaultCloseOptionOnClick}
                >
                  Sign up
                </OptionLink>
              </React.Fragment>
            )}
          </OptionsList>
        </Select>
      )}
    </Wrapper>
  );
};

Component.defaultProps = {
  user: null,
  loading: false,
  error: "",
  signOutRequested: () => {},
};

Component.propTypes = {
  user: PropTypes.object,
  loading: PropTypes.bool.isRequired,
  error: PropTypes.string,
  signOutRequested: PropTypes.func.isRequired,
};

export default Component;
