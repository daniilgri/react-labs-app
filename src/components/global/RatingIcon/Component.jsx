import React, { useMemo } from "react";
import PropTypes from "prop-types";
import StarIcon from "../StarIcon";
import { Wrapper } from "./styles";

const Component = ({ index, rating, hoverRating, onMouseEnter, onMouseLeave, onSaveRating }) => {
  const fill = useMemo(() => {
    if (hoverRating >= index) {
      return "yellow";
    }
    if (!hoverRating && rating >= index) {
      return "yellow";
    }
    return "none";
  }, [rating, hoverRating, index]);

  const handleMouseEnter = () => {
    onMouseEnter(index);
  };

  const handleMouseLeave = () => {
    onMouseLeave();
  };

  const handleIconOnClick = () => {
    onSaveRating(index);
  };

  return (
    <Wrapper
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
      onClick={handleIconOnClick}
    >
      <StarIcon fill={fill} />
    </Wrapper>
  );
};

Component.propTypes = {
  index: PropTypes.number.isRequired,
  rating: PropTypes.number.isRequired,
  hoverRating: PropTypes.number.isRequired,
  onMouseEnter: PropTypes.func.isRequired,
  onMouseLeave: PropTypes.func.isRequired,
  onSaveRating: PropTypes.func.isRequired,
};

export default Component;
