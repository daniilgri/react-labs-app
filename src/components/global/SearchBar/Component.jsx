import React from "react";
import PropTypes from "prop-types";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSearch } from "@fortawesome/free-solid-svg-icons";
import { useFormik } from "formik";

import { Wrapper, InputWrapper, Input, IconWrapper } from "./styles";

const Component = ({ onSet, value }) => {
  const formik = useFormik({
    initialValues: {
      query: value,
    },
    onSubmit: values => {
      onSet(values);
    },
  });

  return (
    <Wrapper onSubmit={formik.handleSubmit}>
      <InputWrapper>
        <Input
          placeholder="Search..."
          name="query"
          id="filmsBoardSearchBar"
          onChange={formik.handleChange}
          value={formik.values.query}
        />
        <IconWrapper type="submit">
          <FontAwesomeIcon icon={faSearch} />
        </IconWrapper>
      </InputWrapper>
    </Wrapper>
  );
};

Component.defaultProps = {
  value: "",
};

Component.propTypes = {
  value: PropTypes.string,
  onSet: PropTypes.func.isRequired,
};

export default Component;
