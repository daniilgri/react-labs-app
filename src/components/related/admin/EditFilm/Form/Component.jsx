import React, { useEffect } from "react";
import { withRouter } from "react-router-dom";
import PropTypes from "prop-types";
import { useFormik } from "formik";

import DescriptionSection from "./DescriptionSection";
import MediaSection from "./MediaSection";
import FormalSection from "./FormalSection";

import { Wrapper, Head, Title, Body, FilledButton, Loading, RedirectLink } from "./styles";

const Component = ({
  openAddScreeningDateModal,
  filmId,
  film,
  editFilmRequested,
  fetchEditFilmRequested,
  loading,
  error,
}) => {
  useEffect(() => {
    fetchEditFilmRequested(filmId);
  }, [fetchEditFilmRequested, filmId]);

  const formik = useFormik({
    initialValues: {
      title: film.title,
      description: film.description,
      ticketPrice: film.ticketPrice,
      imageAsFile: film.image,
      previewImage: film.image,
      tags: film.tags,
      screeningDates: film.screeningDates,
    },
    onSubmit: values => {
      const changedValues = {};
      Object.keys(values).forEach(prop => {
        if (formik.initialValues[prop] !== values[prop]) {
          changedValues[prop] = values[prop];
        }
      });
      if (Object.keys(changedValues).length !== 0 && changedValues.constructor === Object) {
        editFilmRequested({ values, changedValues, filmId: film.id });
      }
    },
    enableReinitialize: true,
  });

  const handleImageUpload = event => {
    event.preventDefault();
    formik.setFieldValue("imageAsFile", event.currentTarget.files[0]);
  };

  const handleScreeningDateAddition = screeningDate => {
    formik.setFieldValue("screeningDates", [...formik.values.screeningDates, screeningDate]);
  };

  const handleScreeningDateDelete = screeningDate => {
    formik.setFieldValue(
      "screeningDates",
      formik.values.screeningDates.filter(sd => sd.date !== screeningDate.date)
    );
  };

  const handleTagAddition = tag => {
    formik.setFieldValue("tags", [...formik.values.tags, tag]);
  };

  const handleTagDelete = tag => {
    formik.setFieldValue(
      "tags",
      formik.values.tags.filter(elTag => elTag !== tag)
    );
  };

  const handleFormOnKeyDown = event => {
    if (event.key === "Enter") {
      event.preventDefault();
    }
  };

  if (loading || error) {
    return <Loading>Loading...</Loading>;
  }

  return (
    <Wrapper onKeyDown={handleFormOnKeyDown} onSubmit={formik.handleSubmit}>
      <Head>
        <Title>Edit film - {film.title}</Title>{" "}
        <RedirectLink to={`/admin/film/${film.id}`}>Go to film card</RedirectLink>
      </Head>
      <Body>
        <DescriptionSection
          title={formik.values.title}
          description={formik.values.description}
          tags={formik.values.tags}
          errors={formik.errors}
          onChange={formik.handleChange}
          onTagAdd={handleTagAddition}
          onTagDelete={handleTagDelete}
        />
        <MediaSection
          errors={formik.errors}
          onImageUpload={handleImageUpload}
          previewImage={formik.values.previewImage}
        />
        <FormalSection
          screeningDates={formik.values.screeningDates}
          ticketPrice={formik.values.ticketPrice}
          errors={formik.errors}
          onChange={formik.handleChange}
          onAddScreeningDateModalOpen={openAddScreeningDateModal}
          onScreeningDateAdd={handleScreeningDateAddition}
          onScreeningDateDelete={handleScreeningDateDelete}
        />
      </Body>

      <FilledButton type="submit">Publish</FilledButton>
    </Wrapper>
  );
};

Component.defaultProps = {
  error: "",
  filmId: "",
};

Component.propTypes = {
  editFilmRequested: PropTypes.func.isRequired,
  openAddScreeningDateModal: PropTypes.func.isRequired,
  filmId: PropTypes.string,
  film: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.bool, PropTypes.number, PropTypes.string, PropTypes.array])
  ).isRequired,
  fetchEditFilmRequested: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
  error: PropTypes.string,
};

export default withRouter(Component);
