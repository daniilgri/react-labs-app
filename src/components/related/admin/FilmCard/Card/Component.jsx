import React, { useEffect } from "react";
import PropTypes from "prop-types";

import FilmBox from "./FilmBox";
import TicketBox from "./TicketBox";
import { Wrapper, Pending } from "./styles";

const Component = ({ filmId, fetchFilmByIdAdminPanelRequested, film, filmError, loading }) => {
  useEffect(() => {
    fetchFilmByIdAdminPanelRequested(filmId);
  }, [fetchFilmByIdAdminPanelRequested, filmId]);

  if (loading || filmError || !film) {
    return <Pending>Loading</Pending>;
  }

  return (
    <Wrapper>
      <FilmBox film={film} />
      <TicketBox screeningDates={film.screeningDates} />
    </Wrapper>
  );
};

Component.defaultProps = {
  filmError: "",
  filmId: "",
  film: {
    title: "",
    description: "",
    tags: [],
    image: "",
    rates: [],
    screeningDates: [],
    ticketPrice: "",
  },
};

Component.propTypes = {
  filmId: PropTypes.string,
  fetchFilmByIdAdminPanelRequested: PropTypes.func.isRequired,
  film: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.bool, PropTypes.number, PropTypes.string, PropTypes.array])
  ),
  filmError: PropTypes.string,
  loading: PropTypes.bool.isRequired,
};

export default Component;
