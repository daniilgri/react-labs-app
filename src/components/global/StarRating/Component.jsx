import React, { useState } from "react";
import PropTypes from "prop-types";
import { v4 as uuidv4 } from "uuid";

import RatingIcon from "../RatingIcon";
import { Wrapper } from "./styles";

const Component = ({ count, rate, onRateSet }) => {
  const [hoverRating, setHoverRating] = useState(0);

  const onMouseEnter = index => {
    setHoverRating(index);
  };

  const onMouseLeave = () => {
    setHoverRating(0);
  };

  const onSaveRating = index => {
    onRateSet(index);
  };

  return (
    <Wrapper>
      {Array.from(Array(count).keys()).map(index => {
        return (
          <RatingIcon
            key={uuidv4()}
            index={index + 1}
            rating={rate}
            hoverRating={hoverRating}
            onMouseEnter={onMouseEnter}
            onMouseLeave={onMouseLeave}
            onSaveRating={onSaveRating}
          />
        );
      })}
    </Wrapper>
  );
};

Component.propTypes = {
  count: PropTypes.number.isRequired,
  rate: PropTypes.number.isRequired,
  onRateSet: PropTypes.func.isRequired,
};

export default Component;
