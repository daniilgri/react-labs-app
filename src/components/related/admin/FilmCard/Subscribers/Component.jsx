import React, { useEffect } from "react";
import PropTypes from "prop-types";
import { v4 as uuidv4 } from "uuid";
import Card from "./Card";

import { Wrapper, Loading, CenterContainer, FetchButton } from "./styles";

const Component = ({
  filmId,
  fetchSubscribersInitialRequested,
  users,
  loading,
  error,
  allCount,
  count,
  fetchSubscribersNextRequested,
}) => {
  const handleGetMoreButtonOnClick = event => {
    event.preventDefault();

    if (count < allCount) {
      fetchSubscribersNextRequested({ filmId });
    }
  };

  useEffect(() => {
    fetchSubscribersInitialRequested({ filmId });
  }, []);

  return (
    <>
      <Wrapper>
        {users.length > 0 && users.map(item => <Card key={uuidv4()} user={item} />)}
      </Wrapper>
      {loading || error ? (
        <Loading>Loading</Loading>
      ) : (
        count < allCount && (
          <CenterContainer>
            <FetchButton onClick={handleGetMoreButtonOnClick}>Get more</FetchButton>
          </CenterContainer>
        )
      )}
    </>
  );
};

Component.defaultProps = {
  users: [],
  error: "",
  count: 0,
  allCount: 0,
  filmId: "",
};

Component.propTypes = {
  filmId: PropTypes.string,
  fetchSubscribersInitialRequested: PropTypes.func.isRequired,
  users: PropTypes.arrayOf(PropTypes.object),
  error: PropTypes.string,
  loading: PropTypes.bool.isRequired,
  count: PropTypes.number,
  allCount: PropTypes.number,
  fetchSubscribersNextRequested: PropTypes.func.isRequired,
};

export default Component;
