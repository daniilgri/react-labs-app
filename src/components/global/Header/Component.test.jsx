import React from "react";
import renderer from "react-test-renderer";
import { BrowserRouter as Router } from "react-router-dom";

import Component from "./Component";

describe("Header component", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Router>
          <Component loading={false} signOutRequested={() => {}} loggedIn={false} />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
