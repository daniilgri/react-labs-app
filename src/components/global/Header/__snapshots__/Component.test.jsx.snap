// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component renders correctly 1`] = `
<nav
  className="sc-bdfBwQ jNAhja"
>
  <div
    className="sc-jSgupP cEWqAA"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-bars fa-w-14 "
      data-icon="bars"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </div>
  <a
    className="sc-gsTCUz jcPZPA"
  >
    Logo
  </a>
  <ul
    className="sc-dlfnbm uDvuY"
  >
    <li
      className="sc-hKgILt iTJVwB"
    >
      <a
        aria-current="page"
        className="sc-eCssSg kJCfdh active"
        href="/"
        onClick={[Function]}
        style={Object {}}
      >
        All films
      </a>
    </li>
  </ul>
  <div
    className="sc-gKsewC domTpR"
  >
    <div
      className="sc-iBPRYJ iZvrwb"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-user-circle fa-w-16 "
        data-icon="user-circle"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 496 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 96c48.6 0 88 39.4 88 88s-39.4 88-88 88-88-39.4-88-88 39.4-88 88-88zm0 344c-58.7 0-111.3-26.6-146.5-68.2 18.8-35.4 55.6-59.8 98.5-59.8 2.4 0 4.8.4 7.1 1.1 13 4.2 26.6 6.9 40.9 6.9 14.3 0 28-2.7 40.9-6.9 2.3-.7 4.7-1.1 7.1-1.1 42.9 0 79.7 24.4 98.5 59.8C359.3 421.4 306.7 448 248 448z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
    <div
      className="sc-fubCfw fiszYb"
    >
      <a
        className="sc-pFZIQ fdxRfJ"
        href="/login"
        onClick={[Function]}
      >
        Login
      </a>
      <a
        className="sc-pFZIQ fdxRfJ"
        href="/signup"
        onClick={[Function]}
      >
        Sign up
      </a>
    </div>
  </div>
</nav>
`;
